src/connection.rs

```diff
@@ -0,0 +1,143 @@
+      = ratio * (S- + d~S1 * d_t1) + (1-ratio) * ~S2
+
+    this implies:
+
+    - the algorithm natively support state sensor(s) with different sampling frequency, incomplete reading,
+    or unreliable reading, as ~S2 variable is merely an optional correction
+    - the interpolation between ~S2 and the first term doesn't need to be linear or additive,
+    e.g. 3D angular interpolation is multiplicative
+    - the ratio can be adjusted based on quality & frequency of state sensor(s)
+
+    most glasses have acc & grav/acc readings in 1 bundle, but I prefer not using this assumption and still update them independently
+
+     */
+    pub const BASE_GRAV_RATIO: f32 = 0.6;
+    pub const BASE_MAG_RATIO: f32 = 0.5;
+}
+
+/*
+high level interface of glasses & state estimation, with the following built-in fusion pipeline:
+
+(the first version should only use complementary filter for simplicity and sanity test)
+
+- roll/pitch <= acc + gyro (complementary filter)
+  - assuming that acc vector always pointed up, spacecraft moving in that direction can create 1G artificial gravity
+    - TODO: this obviously assumes no negative/high G manoeuvre, at which point up d_acc has to be used to get the right up direction
+  - TODO: use ESKF (error-state/multiplicatory KF, https://arxiv.org/abs/1711.02508)
+- gyro-yaw <= gyro (integrate over time)
+- mag-yaw <= mag + roll/pitch (arctan)
+  - TODO: mag calibration?
+     (continuous ellipsoid fitting, assuming homogeneous E-M environment & hardpoint-mounted E-M interference)
+- yaw <= mag-yaw + gyro-gyro (complementary filter)
+  - TODO: use EKF
+*/
+pub struct Connection {
+    pub glasses: Box<dyn ARGlasses>,
+    // estimation
+    pub attitude: UnitQuaternion<f32>,
+    // just old readings
+    // prevAcc: (Vector3<f32>, u64),
+    prev_gyro: (Vector3<f32>, u64),
+    // prevMag: (Vector3<f32>, u64),
+}
+
+pub fn default() -> Result<Connection> {
+    let mut glasses = any_glasses()?;
+    loop {
+        // wait for the first non-zero gyro reading
+        let next_event = glasses.next_event();
+        match next_event {
+            GlassesEvent::AccGyro {
+                accelerometer: _,
+                gyroscope,
+                timestamp,
+            } => {
+                if gyroscope != Vector3::zeros() {
+                    let conn = Connection {
+                        glasses,
+                        attitude: UnitQuaternion::identity(),
+                        prev_gyro: (gyroscope, timestamp),
+                    };
+                    return Ok(conn);
+                }
+            }
+            _ => {}
+        }
+    }
+}
+
+impl Connection {
+    const UP: Vector3<f32> = Vector3::new(0.0, 1.0, 0.0);
+    const NORTH: Vector3<f32> = Vector3::new(1.0, 0.0, 0.0);
+
+    fn disconnect(self) -> () {
+        self.glasses.deref();
+    }
+
+    fn update_gyro(&mut self, gyro: Vector3<f32>, t: u64) -> () {
+        let d_t1 = ((t - self.prev_gyro.1) as f32) / 1000.0;
```

Comment:
What is `1000.0` here?

Reply:
milliseconds? Sorry I should read the manual :face_with_spiral_eyes:

Reply:
yeah the speed is rad/sec, while timestamps are milliseconds, this should be consistent

Reply:
This timestamp is in microseconds. Please see the docstring on the AccGyro enum value.

Did you test this? Did it work?